### DISCUSSION TITLE
Model Tuning Code

### DISCUSSION THREAD

**Discussion by eeinnew** [teamalgoritma/community#127] [Q&A] [answered]
Hello, would like to ask about this `naive bayes` model tuning  code from this rpubs : https://rpubs.com/nabiilahardini/wine-ndf .I do not understand what this code stands for
```
library(plotly)

# model tuning - metrics function
metrics <- function(cutoff, prob, ref, postarget, negtarget) 
{
  predict <- factor(ifelse(prob >= cutoff, postarget, negtarget))
  conf <- caret::confusionMatrix(predict , ref, positive = postarget)
  acc <- conf$overall[1]
  rec <- conf$byClass[1]
  prec <- conf$byClass[3]
  spec <- conf$byClass[2]
  mat <- t(as.matrix(c(rec , acc , prec, spec))) 
  colnames(mat) <- c("recall", "accuracy", "precicion", "specificity")
  return(mat)
}
co <- seq(0.01,0.99,length=100)
result <- matrix(0,100,4)

# apply function metrics
for(i in 1:100){
  result[i,] = metrics(cutoff = co[i], 
                     prob = naive_table$quality_eprob, 
                     ref = as.factor(ifelse(naive_table$quality == "Excellent", 1, 0)), 
                     postarget = "1", 
                     negtarget = "0")
}

# visualize
ggplotly(tibble("Recall" = result[,1],
           "Accuracy" = result[,2],
           "Precision" = result[,3],
           "Specificity" = result[,4],
                   "Cutoff" = co) %>% 
  gather(key = "Metrics", value = "value", 1:4) %>% 
  ggplot(aes(x = Cutoff, y = value, col = Metrics)) +
  geom_line(lwd = 1.5) +
  scale_color_manual(values = c("darkred","darkgreen","orange", "blue")) +
  scale_y_continuous(breaks = seq(0,1,0.1), limits = c(0,1)) +
  scale_x_continuous(breaks = seq(0,1,0.1)) +
  labs(title = "Tradeoff Model Perfomance") +
  theme_minimal() +
  theme(legend.position = "top",
        panel.grid.minor.y = element_blank(),
        panel.grid.minor.x = element_blank()))
```

**Comment by NabiilahArdini** [comment, marked as answer]
Hello @eeinnew, 

Those are codes for making **the visualization for 4 metrics (Accuracy, Recall, Precision, Specificity) for each of the threshold that used**. Using the visualization, we can pick which threshold that satisfy our need on model performance. This code has also later developed into a package that you can directly use here: https://github.com/ahmadhusain/cmplot/

Still, I hope this explanation can help you understand the code better:
1. Making metrics function. This function will predict and create confusion matrix, then save the result of the 4 metrics for every threshold used. This function will be used in the later step. 
```
metrics <- function(cutoff, prob, ref, postarget, negtarget) 
{
  predict <- factor(ifelse(prob >= cutoff, postarget, negtarget))
  conf <- caret::confusionMatrix(predict , ref, positive = postarget)
  acc <- conf$overall[1]
  rec <- conf$byClass[1]
  prec <- conf$byClass[3]
  spec <- conf$byClass[2]
  mat <- t(as.matrix(c(rec , acc , prec, spec))) 
  colnames(mat) <- c("recall", "accuracy", "precicion", "specificity")
  return(mat)
}
```

2. Prepare the cut off / threshold value. Here we make a sequence from 0.01 - 0.99 (100 points of cut off). This will be used in the later step. 
```
co <- seq(0.01,0.99,length=100)
```

3. Prepare empty matrix (0 value, for 100 x 4 matrix), for later be used to store the 4 metrics. We make 100 rows for each threshold and 4 column for each metrics.
``` 
result <- matrix(0,100,4)
```

4. Executing metrics function in a for loop. For every i (index) from 1:100, from `co` (threshold object), execute the metrics function, and save the result in each rows of the object `result`.

```
for(i in 1:100){
  result[i,] = metrics(cutoff = co[i], 
                     prob = naive_table$quality_eprob, 
                     ref = as.factor(ifelse(naive_table$quality == "Excellent", 1, 0)), 
                     postarget = "1", 
                     negtarget = "0")
}
```
5. Visualize the result object into an interactive plot

```
ggplotly(tibble("Recall" = result[,1],
           "Accuracy" = result[,2],
           "Precision" = result[,3],
           "Specificity" = result[,4],
                   "Cutoff" = co) %>% 
  gather(key = "Metrics", value = "value", 1:4) %>% 
  ggplot(aes(x = Cutoff, y = value, col = Metrics)) +
  geom_line(lwd = 1.5) +
  scale_color_manual(values = c("darkred","darkgreen","orange", "blue")) +
  scale_y_continuous(breaks = seq(0,1,0.1), limits = c(0,1)) +
  scale_x_continuous(breaks = seq(0,1,0.1)) +
  labs(title = "Tradeoff Model Perfomance") +
  theme_minimal() +
  theme(legend.position = "top",
        panel.grid.minor.y = element_blank(),
        panel.grid.minor.x = element_blank()))
```

You may explore each of the function a little bit more if you are interested. 
I hope this answer can help you. If it do, please mark this as an answer. Thank you! :)

**Comment by inytss** [comment]
Halo @eeinnew ,
Untuk code tersebut memunculkan metric recall, precision, akurasi, dan specificity.
Seperti penjelasakn yang telah diberikan oleh Mbak @NabiilahArdini  pada jawaban sebelumnya, code tersebut digunakan untuk melakukan tunning threshold yang akan digunakan dalam menentukan kelas target karena default threshold yang digunakan adalah 0.5. Ketika probability > 0.5 akan masuk kelas 1, jika tidak akan masuk kelas 0.

Berdasarkan plot yang dimunculkan, Ibu bisa memilih menggunakan threshold berapa agar metric yang ingin Ibu dapatkan menjadi tinggi dengan melakukan hover kursor pada plot tersebut.

Sebagai contoh berikut ini, pada cut off/threshold 0.78, model yang kita miliki bisa mendapatkan metric akurasi = 75%, recall = 74%, specificity = 76%, dan precision = 46%. Jika merasa dengan threshold tersebut masih kurang optimal performa modelnya, Ibu bisa coba mentuning lagi dan melihat pada threshold berapakah kita akan membedakan kelas target kita.

![Screenshot from 2022-06-03 15-38-59](https://user-images.githubusercontent.com/49224928/171820207-7ac2d447-cea7-45ad-bc4d-44307f0fdeea.png)
